BZ2039509: Adds the instruction to read the pod logs (#48549)
openshift/openshift-docs: modules/nw-metallb-software-components.adoc

[id="nw-metallb-software-components_{context}"]
= MetalLB software components

When you install the MetalLB Operator, the `metallb-operator-controller-manager` deployment starts a pod. The pod is the implementation of the Operator. The pod monitors for changes to all the relevant resources.

When the Operator starts an instance of MetalLB, it starts a `controller` deployment and a `speaker` daemon set.

`controller`::
The Operator starts the deployment and a single pod. When you add a service of type `LoadBalancer`, Kubernetes uses the `controller` to allocate an IP address from an address pool.

> In case of a service failure, ensure you have the following entry in your `controller` pod logs:
> +
> [source,terminal]

Reviewer: ```suggestion
In case of a service failure, verify you have the following entry in your `controller` pod logs:
+
.Example output
[source,terminal]
```
Just a small nit. See the [repo guidelines ](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output) for adding an label to an example output. Also, _ensure_ is not wrong, but I think in our docs it often implies, "make sure you did this thing." To me, _verify_ is a bit clearer in this case. WDYT?

Author: Agreed. Verify sounds more better

Author: addressed

Author: Addressed

----
"event":"ipAllocated","ip":"172.22.0.201","msg":"IP address assigned by controller
----

`speaker`::

> The Operator starts a daemon set for `speaker` pods. By default, a pod is started on each node in your cluster. You can limit the pods to specific nodes by specifying a node selector in the `MetalLB` custom resource when you start MetalLB. If the `controller` allocated the ip address to the service and service is still unavailable, read the `speaker` pod logs. If the speaker pod is unavailable, run the `oc describe pod -n` command.

Reviewer 2: ```suggestion
The Operator starts a daemon set for `speaker` pods. By default, a pod is started on each node in your cluster. You can limit the pods to specific nodes by specifying a node selector in the `MetalLB` custom resource when you start MetalLB. If the `controller` allocated the IP address to the service and service is still unavailable, read the `speaker` pod logs. If the speaker pod is unavailable, run the `oc describe pod -n` command.
```

Author: Addressed

Author: Added the section name with the link

> The Operator starts a daemon set for `speaker` pods. By default, a pod is started on each node in your cluster. You can limit the pods to specific nodes by specifying a node selector in the `MetalLB` custom resource when you start MetalLB. If the `controller` allocated the IP address to the service and service is still unavailable, read the `speaker` pod logs. If the speaker pod is unavailable, run the `oc describe pod -n` command.

Reviewer: ```suggestion
The Operator starts a daemon set for `speaker` pods. By default, a pod is started on each node in your cluster. You can limit the pods to specific nodes by specifying a node selector in the `MetalLB` custom resource when you start MetalLB. If the `controller` allocated the IP address to the service and service is still unavailable, read the `speaker` pod logs. If the `speaker` pod is unavailable, run the `oc describe pod -n` command.
```

Author: Addressed

+
For layer 2 mode, after the `controller` allocates an IP address for the service, the `speaker` pods use an algorithm to determine which `speaker` pod on which node will announce the load balancer IP address.

> The algorithm involves hashing the node name and the load balancer IP address. See the section about external traffic policy for more information.

Reviewer 2: Should the name of the section also be added for clarification in the second sentence? Up to you though! 😀

Author: Addressed

Author: Added the section name with the link

Reviewer 2: @xenolinux Hi my apologies, I should've been more clear! The link won't work becuase we don't put xrefs in a module. I mean't adding the name of the section, in this case like, 
> See "MetalLB` and external traffic policy" for more information.
 instead of describing it

Author: Addressed

> The algorithm involves hashing the node name and the load balancer IP address. See the section MetalLB and external traffic policy for more information.

Reviewer: ```suggestion
The algorithm involves hashing the node name and the load balancer IP address. For more information, see "MetalLB and external traffic policy".
```
Putting the section in quotes will help the customer understand what you are referring to.

Author: Addressed

// IETF treats protocol names as proper nouns.
The `speaker` uses Address Resolution Protocol (ARP) to announce IPv4 addresses and Neighbor Discovery Protocol (NDP) to announce IPv6 addresses.

> For BGP mode, after the `controller` allocates an IP address for the service, each `speaker` pod advertises the load balancer IP address with its BGP peers. You can configure which nodes start BGP sessions with BGP peers.

Reviewer 2: Ignore if I'm missing something here! Is BGM an abbreviation, and if so is it already defined?

Author: Do you mean BGP? It seems like BGP wasn't defined before so I added the long form.

Author: Addressed

`controller`::
The Operator starts the deployment and a single pod.
When you add a service of type `LoadBalancer`, Kubernetes uses the `controller` to allocate an IP address from an address pool.

> Ensure you have the following entry in your `controller` pod logs:

Reviewer 3: @xenolinux Could you amend this statement as "Ensure you have the following entry in your controller pod logs and in case of failure error related to the service".

Controller is responsible for address assignment and speaker is responsible for IP address announcement, if the service does not have IP address customer needs to look at controller logs, if ip address is assigned but service is not working then speaker logs need to examined.

Author: Addressed

+
[source,terminal]
----
"event":"ipAllocated","ip":"172.22.0.201","msg":"IP address assigned by controller
----

`speaker`::
The Operator starts a daemon set for `speaker` pods.
By default, a pod is started on each node in your cluster.
You can limit the pods to specific nodes by specifying a node selector in the `MetalLB` custom resource when you start MetalLB.

> If a `speaker` pod becomes unavailable, read the `speaker` pod logs.

Reviewer 3: It is not when 'speaker' pod becomes unavailable customer should read to speaker logs. It should be "In case of failure read the logs speaker pod. If the speaker pod itself is not in running state, oc describe pod <pod name> -n <namespace> can give more details".
The likelihood of speaker pod being unavailable is very less.

Author: Addressed

Author: @asood-rh Could you PTAL?
